### PR TITLE
fix: Azure fragment store with multiple tenants

### PR DESCRIPTION
**Pull request by jshearer**
I knew that the service credentials would be scoped to tenant IDs, but I did not realize that the `service.Client`s and `UserDelegationCredential`s would also be tenant-scoped. This was causing problems where the credentials of whatever tenant was first seen would be used to sign _all_ `SignGet` URLs, resulting in signature mismatch errors when the blob was owned by a different tenant.

This updates the cache logic to keep track of both the service clients and the user delegation credentials by tenant ID, which should solve the problem.

### CODE REVIEW

File: broker/fragment/store_azure.go [hunk 1]
```diff
@@ -222,23 +237,100 @@ func getAzureStorageCredential(coreCredential azcore.TokenCredential, tenant str
...
-		cfg.accountTenantID, cfg.storageAccountName, cfg.containerName, cfg.prefix = ep.Host, splitPath[0], splitPath[1], strings.Join(splitPath[2:], "/")
+		cfg.containerName, cfg.prefix = storageMapping.Host, storageMapping.Path[1:]
+	} else if storageMapping.Scheme == "azure-ad" {
+		cfg.accountTenantID, cfg.storageAccountName, cfg.containerName, cfg.prefix = storageMapping.Host, splitPath[0], splitPath[1], strings.Join(splitPath[2:], "/")
+	}
+
+	return cfg, nil
+}
+
+func (a *azureBackend) getAzureServiceClient(storageMapping *url.URL) (client *service.Client, err error) {
```
Comment (jgraettinger) [inline broker/fragment/store_azure.go:263]: this needs to be guarded by a mutex. you're manipulating `a` here.

File: broker/fragment/store_azure.go [hunk 2]
```diff
@@ -324,7 +403,16 @@ func (a *azureBackend) buildBlobURL(cfg AzureStoreConfig, client pipeline.Pipeli
...
-func (a *azureBackend) getUserDelegationCredential() (*service.UserDelegationCredential, error) {
+func (a *azureBackend) getUserDelegationCredential(storageMapping *url.URL) (*service.UserDelegationCredential, error) {
+	if a.udcs == nil {
+		a.udcs = make(map[string]UdcAndExp)
+	}
+	var cfg, err = parseAzureStorageMapping(storageMapping)
+	if err != nil {
+		return nil, err
+	}
+	var udc, hasCachedUdc = a.udcs[cfg.accountTenantID]
```
Comment (jgraettinger) [inline broker/fragment/store_azure.go:414]: This, and the initialization above, need to be guarded by a mutex.

Please carefully review everywhere that you're mutating and accessing these maps. It always always needs to be guarded.

I would recommend wrapping it in a brief critical section here which handles initialization and pulls out `udc` & `hasCachedUdc`. Then, below where you upsert, lock there as well.

I would rename `clientMu` to just `mu`, and essentially any time you're touching state attached to `a`, it needs to be within the scope of a briefly-held lock. It's perfectly fine, i'd think, if multiple goroutines each independently fetch credentials and overwrite each other :shrug: so you don't need to hold locks for long.

**Review by jgraettinger — CHANGES_REQUESTED** [review]

File: broker/fragment/store_azure.go [hunk 3]
```diff
@@ -56,14 +64,14 @@ func (a *azureBackend) Provider() string {
 
 // See here for an example of how to use the Azure client libraries to create signatures:
 // https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/storage/azblob/service/examples_test.go#L285
-func (a *azureBackend) SignGet(ep *url.URL, fragment pb.Fragment, d time.Duration) (string, error) {
-	cfg, _, err := a.azureClient(ep)
+func (a *azureBackend) SignGet(storageMapping *url.URL, fragment pb.Fragment, d time.Duration) (string, error) {
```
[the remark this answers is not in the document]
Comment (jshearer) [reply broker/fragment/store_azure.go:67]: kk, I'll go with `endpoint`. Even though I know it's a convention, the short variable names make it a lot harder to understand what anything is IMO

File: broker/fragment/store_azure.go (hunk 2 above)
Comment (jshearer) [reply broker/fragment/store_azure.go:414]: Okay, we talked on a VC and my understanding here is: any area that reads _or mutates_ state, and can be entered from multiple threads/goroutines needs a critical section. I believe I got all of them here.

File: broker/fragment/store_azure.go [hunk 4]
```diff
@@ -222,47 +237,133 @@ func getAzureStorageCredential(coreCredential azcore.TokenCredential, tenant str
...
-		return
+func (a *azureBackend) getAzureServiceClient(endpoint *url.URL) (client *service.Client, err error) {
+	var cfg azureStoreConfig
+
+	if cfg, err = parseAzureEndpoint(endpoint); err != nil {
+		return nil, err
 	}
 
-	var credentials azblob.Credential
+	a.mu.Lock()
```
Comment (jgraettinger) [inline broker/fragment/store_azure.go:270]: nit: can you arrange so that any constructed instance of `azureBackend` already has all maps initialized? Then you don't need to worry about their being nil.

For example:
```
&azureBackend {
   foo: make(map[string]*service.Client),
   bar: make(map[string]string),
}
```

then you can omit nil checks everywhere.

File: broker/fragment/store_azure.go [hunk 5]
```diff
@@ -222,47 +237,133 @@ func getAzureStorageCredential(coreCredential azcore.TokenCredential, tenant str
...
 		}
 		serviceClient, err := service.NewClientWithSharedKeyCredential(cfg.serviceUrl(), sharedKeyCred, &service.ClientOptions{})
 		if err != nil {
-			return cfg, nil, err
+			return nil, err
 		}
-		a.svcClient = *serviceClient
+
+		a.mu.Lock()
+		a.svcClients[accountName] = *serviceClient
```
Comment (jgraettinger) [inline broker/fragment/store_azure.go:301]: You're dereferencing (making a shallow copy) here and I don't understand why. Seems like `svcClients` should be holding pointers, since that's what's returned by service.NewClientWithSharedKeyCredential().

File: broker/fragment/store_azure.go [hunk 6]
```diff
@@ -34,20 +35,27 @@ type AzureStoreConfig struct {
...
+}
+
 type azureBackend struct {
-	clients   map[string]pipeline.Pipeline
-	svcClient service.Client
-	clientMu  sync.Mutex
-	udc       *service.UserDelegationCredential
-	udcExp    *time.Time
+	// This is a cache of configured Pipelines for each tenant. These do not expire
+	clients map[string]pipeline.Pipeline
```
Comment (jgraettinger) [inline broker/fragment/store_azure.go:53]: nit: can these be called `pipelines` and `services` perhaps, here and elsewhere? Or `pipelines` and `clients` ?

It's hard to follow which is which.

File: broker/fragment/store_azure.go (hunk 5 above)
Comment (jshearer) [reply broker/fragment/store_azure.go]: Fixed, good point about `svcClients` (now just `clients`) holding a map of pointers.

File: broker/fragment/store_azure.go (hunk 4 above)
Comment (jshearer) [reply broker/fragment/store_azure.go]: Done! It was only constructed in one place, so easy to update.
